airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialLoadGlobalStateManager.java

```diff
@@ -0,0 +1,132 @@
+  private final Map<AirbyteStreamNameNamespacePair, OrderedColumnInfo> pairToOrderedColInfo;
+  private final CdcState cdcState;
+
+  // Only one global state is emitted, which is fanned out into many entries in the DB by platform. As
+  // a result, we need to keep track of streams that
+  // have completed the snapshot.
+  private final Set<AirbyteStreamNameNamespacePair> streamsThatHaveCompletedSnapshot;
+  public MssqlInitialLoadGlobalStateManager(final InitialLoadStreams initialLoadStreams,
+      final Map<AirbyteStreamNameNamespacePair, OrderedColumnInfo> pairToOrderedColInfo,
+      final CdcState cdcState,
+      final ConfiguredAirbyteCatalog catalog) {
+    this.cdcState = cdcState;
+    this.pairToOrderedColLoadStatus = MssqlInitialLoadStateManager.initPairToOrderedColumnLoadStatusMap(initialLoadStreams.pairToInitialLoadStatus());
+    this.pairToOrderedColInfo = pairToOrderedColInfo;
+    this.streamsThatHaveCompletedSnapshot = initStreamsCompletedSnapshot(initialLoadStreams, catalog);
+  }
+
+  private static Set<AirbyteStreamNameNamespacePair> initStreamsCompletedSnapshot(final InitialLoadStreams initialLoadStreams,
+      final ConfiguredAirbyteCatalog catalog) {
+    final Set<AirbyteStreamNameNamespacePair> streamsThatHaveCompletedSnapshot = new HashSet<>();
+    catalog.getStreams().forEach(configuredAirbyteStream -> {
+      if (!initialLoadStreams.streamsForInitialLoad().contains(configuredAirbyteStream)
+          && configuredAirbyteStream.getSyncMode() == SyncMode.INCREMENTAL) {
+        streamsThatHaveCompletedSnapshot.add(
+            new AirbyteStreamNameNamespacePair(configuredAirbyteStream.getStream().getName(), configuredAirbyteStream.getStream().getNamespace()));
+      }
+    });
+    return streamsThatHaveCompletedSnapshot;
+  }
+
+  @Override
+  public AirbyteStateMessage createIntermediateStateMessage(final AirbyteStreamNameNamespacePair pair, final OrderedColumnLoadStatus ocLoadStatus) {
+    final List<AirbyteStreamState> streamStates = new ArrayList<>();
+    streamsThatHaveCompletedSnapshot.forEach(stream -> {
+      final DbStreamState state = getFinalState(stream);
+      streamStates.add(getAirbyteStreamState(stream, Jsons.jsonNode(state)));
+    });
+    streamStates.add(getAirbyteStreamState(pair, (Jsons.jsonNode(ocLoadStatus))));
+    final AirbyteGlobalState globalState = new AirbyteGlobalState();
+    globalState.setSharedState(Jsons.jsonNode(cdcState));
+    globalState.setStreamStates(streamStates);
+
+    return new AirbyteStateMessage()
+        .withType(AirbyteStateType.GLOBAL)
+        .withGlobal(globalState);
+  }
+
+  private AirbyteStreamState getAirbyteStreamState(final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair pair, final JsonNode stateData) {
+    assert Objects.nonNull(pair);
+    assert Objects.nonNull(pair.getName());
+    assert Objects.nonNull(pair.getNamespace());
```

Comment:
How useful are these assertions? They will be ignored outside of tests. In tests, an NPE will have the same effect.

Reply:
We really should start using the @NotNull annotations and let our static tools warn us when we're mucking things up